### PR TITLE
FileResponse: support for special characters in fileName  (+ suppress warning when nullable is enabled in C# clients) 

### PR DESCRIPTION
**Pull request by knoepdan**
**Warning suppressed**
When the Nullable feature is enabled and Warning are treated as errors, the generated C# client will not compile due to the fact, that deserialization might fail and newtonsoft will the return null. Therefore, this pull request will suppress the corresponding warning. 

Example code snippet that might cause an error depending on the C# projec settings: 
`      
  public static Path FromJson(string data)
        {
            return Newtonsoft.Json.JsonConvert.DeserializeObject<Path>(data, new Newtonsoft.Json.JsonSerializerSettings());
        }`

Remark: adding the ? to mark the return value as nullable would have been possible as well. However, this might not be compatible with older C# versions, so i guess its safer to just suppress the error.

### CODE REVIEW

**Comment by knoepdan** [comment]
:-(  oh no.. i didnt mean the second commit to be part of this pull request. That should have been a separate one.   (It's a long long time since i touched GitHub again.. nevertheless, i should have known). 

This commit fixes an issue with the fileResponse when the fileName contains special characters (such as Umlaute).


To support special characters in a filename in asp.net, one can use the ContentDisposition.FileNameStar property (instead of FileName). Example in the C# code below (which shows usage of both properties)

`    
            var response = new HttpResponseMessage(HttpStatusCode.OK);
            byte[] contents = GetFileResponse(out string filename);

            response.Content = new ByteArrayContent(contents);
            response.Content.Headers.ContentDisposition = new ContentDispositionHeaderValue("attachment");

            // fileresponse without support for special characters (e.g.: Umlaute) 
            //response.Content.Headers.ContentDisposition.FileName = filename; // filename="=?utf-8?B?w6TDpMOkMjAyMjAxMDZfMTc0ODUzLkNTVg==?="  (incorrect with special characters)

            // or with support for special characters (e.g. Umlaute)
            response.Content.Headers.ContentDisposition.FileNameStar = filename; // filename*=utf-8''%C3%A4%C3%A4%C3%A420220106_174853.CSV

            response.Content.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");      
`

**Comment by knoepdan** [comment]
Actually, i meant to do 2 commits, one for the warning and one for the fileResponse enhancement.     (It's a long long time since i touched GitHub again.. nevertheless, i should have known.). I hope its still ok like this since one commit is just about a warning.

**Comment by RicoSuter** [comment]
Lgtm, thanks for the PR!

**Comment by KostaMadorsky** [comment]
Looks like this PR broke working parsing in my case: #3865

**Comment by RicoSuter** [comment]
@KostaMadorsky @knoepdan Can you provide a fix for this so that it's not introducing a regression and also supports this new scenario? Otherwise I have to rollback this change.

**Comment by KostaMadorsky** [comment]
@RicoSuter I would not be able to push a well-tested fix for this any time soon, unfortunately.
I think the proper version should come with a bunch of unit tests (the new regex does not look straightforward).

For now, something like that would do (the regex is the same as in the new version):

```
const contentDisposition = `attachment; filename="Certificate - Test questions - 2022-02-07_00-10.pdf"; filename*=UTF-8''Certificate%20-%20Test%20questions%20-%202022-02-07_00-10.pdf`;
const fileNameRgx = /filename[^;=\n]*=(?:(\\?['"])(.*?)\1|(?:[^\s]+'.*?')?([^;\n]*))/g;

let fileName;
let matchArray;

while ((matchArray = fileNameRgx.exec(contentDisposition)) != null) {
  if (matchArray.length < 4) {
    continue;
  }

  fileName = matchArray[3] || matchArray[2] || fileName;
  console.log(fileName);
}

if (fileName) {
  fileName = decodeURIComponent(fileName);
}

console.log('final: ', fileName);
```

image


https://onecompiler.com/javascript/3xspf4sbx
image

**Comment by knoepdan** [comment]
Sorry for the problems. That should not have happened.  (ohh i actually hate Regex, so hard to understand sometimes)
I also cannot provide a well tested fix so quickly. And yes unit tests would be nice. 

However, what I propose as a quick fix that certainly works is the following: 


`let fileNameMatch = contentDisposition ? /filename="?([^"]*?)"?(;|$)/g.exec(contentDisposition) : undefined;
let fileName = fileNameMatch && fileNameMatch.length > 1 ? fileNameMatch[1] : undefined;
if(!fileName){
	fileNameMatch = contentDisposition ? /filename[^;=\n]*=(?:(\\?['"])(.*?)\1|(?:[^\s]+'.*?')?([^;\n]*))/g.exec(contentDisposition) : undefined;
	fileName = fileNameMatch && fileNameMatch.length > 1 ? decodeURIComponent(fileNameMatch[fileNameMatch.length - 1]) : undefined;
}`

Basically, i use the old Regex which has been well tested in the field, and if it returns undefined, we try with the new one. 
However, i don't think this should be the final solution. (and even if it was, there should be unit tests etc.). Will still create a pull request for this as I think its a) better than the current solution and b) better than the previous state.

**Comment by KostaMadorsky** [comment]
@knoepdan it will not use UTF8 filename if there is non-UTF8  filename in the string with your solution.
Could you try mine and see if it works?
Tried 4 different scenarios on the link above.

**Comment by KostaMadorsky** [comment]
@knoepdan your regex always returns 4 array elements if matched: whole match string and 3 capturing groups.
So then you need to run it until there are matches (while loop in my code) and try to pull values from the 3rd capturing group
If no UTF8 filename, it will fall back to the second capturing group, which works for non-UTF8 filename.

**Comment by RicoSuter** [comment]
Reverted this PR => v13.15.8
Please find a solution which works for everyone in the other PR...
